cpp/src/strings/split/split_re.cu

```diff
@@ -0,0 +1,374 @@
+
+  // convert match counts to token offsets
+  auto map_fn = [d_strings, d_offsets, max_tokens] __device__(auto idx) {
+    return d_strings.is_null(idx) ? 0 : std::min(d_offsets[idx], max_tokens) + 1;
+  };
+  thrust::transform_exclusive_scan(
+    rmm::exec_policy(stream), begin, end + 1, d_offsets, map_fn, 0, thrust::plus<offset_type>{});
+
+  // the last offset entry is the total number of tokens to be generated
+  auto const total_tokens = cudf::detail::get_value<offset_type>(offsets, strings_count, stream);
+
+  // generate tokens for each string
+  rmm::device_uvector<string_index_pair> tokens(total_tokens, stream);
+  auto const regex_insts = d_prog.insts_counts();
+  if (regex_insts <= RX_SMALL_INSTS) {
+    token_reader_fn<RX_STACK_SMALL> reader{d_strings, d_prog, direction, d_offsets, tokens.data()};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, reader);
+  } else if (regex_insts <= RX_MEDIUM_INSTS) {
+    token_reader_fn<RX_STACK_MEDIUM> reader{d_strings, d_prog, direction, d_offsets, tokens.data()};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, reader);
+  } else if (regex_insts <= RX_LARGE_INSTS) {
+    token_reader_fn<RX_STACK_LARGE> reader{d_strings, d_prog, direction, d_offsets, tokens.data()};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, reader);
+  } else {
+    token_reader_fn<RX_STACK_ANY> reader{d_strings, d_prog, direction, d_offsets, tokens.data()};
+    thrust::for_each_n(rmm::exec_policy(stream), begin, strings_count, reader);
+  }
+
+  return tokens;
+}
+
+/**
+ * @brief Returns string pair for the specified column for each string in `d_strings`
+ *
+ * This is used to build the table result of a split.
+ * Null is returned if the row is null or if the `column_index` is larger
+ * than the token count for that string.
+ */
+struct tokens_transform_fn {
+  column_device_view const d_strings;
+  string_index_pair const* d_tokens;
+  offset_type const* d_token_offsets;
+  size_type const column_index;
+
+  __device__ string_index_pair operator()(size_type idx) const
+  {
+    auto const offset      = d_token_offsets[idx];
+    auto const token_count = d_token_offsets[idx + 1] - offset;
+    return (column_index > token_count - 1) || d_strings.is_null(idx)
+             ? string_index_pair{nullptr, 0}
+             : d_tokens[offset + column_index];
+  }
+};
+
+std::unique_ptr<table> split_re(strings_column_view const& input,
+                                std::string const& pattern,
+                                split_direction direction,
+                                size_type maxsplit,
+                                rmm::cuda_stream_view stream,
+                                rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(!pattern.empty(), "Parameter pattern must not be empty");
+
+  auto const max_tokens    = maxsplit > 0 ? maxsplit : std::numeric_limits<size_type>::max();
```

Comment:
This line could be moved into the `generate_tokens` function since it is repeated by both calling functions and is needed for correctness.